### PR TITLE
NIFI-9701 - Updated the ListSFTP NO_TRACKING listing strategy to crea…

### PR DESCRIPTION
**Pull request by thenatog**
…te a single flow file when using a record writer.

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java [hunk 1]
```diff
@@ -74,12 +82,14 @@ public void setUp() throws Exception {
...
     public void tearDown() throws Exception {
         sshServer.stopServer();
+        Files.deleteIfExists(Paths.get(sshServer.getVirtualFileSystemPath()));
     }
 
     @Test
-    public void testRunFileFound() {
+    public void testRunFileFound() throws InterruptedException {
+        writeTempFile(1);
+        Thread.sleep(STABILITY_WAIT);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java:91]: It would be helpful to track down potential issues as opposed to introducing `Thread.sleep()`, which creates more test performance problems.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java [hunk 2]
```diff
@@ -93,7 +103,61 @@ public void testRunFileFound() {
...
+    }
+
+    @Test
+    public void testRunWithRecordWriter() throws InitializationException, InterruptedException {
+        writeTempFile(3);
+        RecordSetWriterFactory recordWriter = getCsvRecordWriter();
+        runner.addControllerService("csv-record-writer", recordWriter);
+        runner.setProperty(AbstractListProcessor.RECORD_WRITER, "csv-record-writer");
+        runner.enableControllerService(recordWriter);
+        Thread.sleep(STABILITY_WAIT);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java:115]: Perhaps calling `runner.assertValid()` here would be helpful?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestDeduplicateRecord.java [hunk 3]
```diff
@@ -271,7 +271,7 @@ void doCountTests(int failure, int original, int duplicates, int notDuplicates,
         }
     }
 
-    private static final class MockCacheService<K, V> extends AbstractControllerService implements DistributedMapCacheClient {
+    static final class MockCacheService<K, V> extends AbstractControllerService implements DistributedMapCacheClient {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestDeduplicateRecord.java:274]: Is this change necessary as part of this PR?

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for addressing this issue @thenatog! The change to `AbstractListProcessor` seems straightforward, but the introduction of `Thread.sleep()` in `TestListSFTP` seems problematic. It would be helpful to track down the source of the issue that appears to need more time for completion. Perhaps it would be better to introduce tests for `AbstractListProcessor` instead of `ListSFTP` if that is the source of the problem?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestDeduplicateRecord.java (hunk 3 above)
Comment (thenatog) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestDeduplicateRecord.java:274]: I did use it in my tests in order to test the listing "AbstractListProcessor.BY_ENTITIES" option. I could try moving the MockCacheService to a separate class that both test classes could use?

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestDeduplicateRecord.java:274]: Yes, moving it to a shared class seems like the best approach.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java (hunk 1 above)
Comment (thenatog) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java:91]: I've been trying to figure out what I can do about this. The issue is that I will intermittently see '        runner.assertTransferCount(ListSFTP.REL_SUCCESS, 1);' fail with a result of 0.

I've tried:
- Asserting that the file count available in the virtual SSH/SFTP server matches the number created for the test (the files are present to be listed)
- Asserting that the runner is valid
- Asserting that the controller services are valid

I've currently got it working, rather than using a thread sleep, by setting the runner's scheduler to 500ms instead, and running 2 iterations... If that's not acceptable, I may need to perform the test in TestAbstractListProcessor. However it seems like it could be useful to increase the testing coverage for TestListSFTP.

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java:91]: Thanks for the update. Running multiple iterations might work, although that seems questionable.

Since the listing strategy handling is not part of `ListSFTP` itself, it seems like these tests would be better run in the context of the class that implements the logic. Not sure if that just shifts the problem to a different location, but it might be worth considering.

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java:91]: The `ListSFTP` processor includes a file filter that checks the minimum age based on the current system time in milliseconds. The `SFTPTransfer.newFileInfo()` method sets the last modified time, but apparently it only provides resolution to the second and includes a multiplication by 1000 to get milliseconds. With the default minimum age set to zero seconds, this might explain the reason that the test sometimes misses files.

With that background, another solution may be to set the minimum age to a negative number.

Comment (thenatog) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestListSFTP.java:91]: Yes I think you're correct, I think the file age was the key problem. I'll submit an update to the PR in a few moments.

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the test adjustments @thenatog, the current version looks good! +1 merging.
